tensorpack/dataflow/imgaug/geometry.py

```diff
@@ -150,3 +150,98 @@ def largest_rotated_rect(w, h, angle):
+
+    def __init__(self, scale=None, translate_frac=None, rotate_max_deg=0.0, shear=0.0,
+                 interp=cv2.INTER_LINEAR, border=cv2.BORDER_REPLICATE, border_value=0):
+        """
+        Args:
+            scale (tuple of 2 floats): scaling factor interval, e.g (a, b), then scale is
+                randomly sampled from the range a <= scale <= b
+            translate_frac (tuple of 2 floats): tuple of max abs fraction for horizontal
+                and vertical shifts. For example translate_frac=(a, b), then horizontal shift
+                is randomly sampled in the range 0 < dx < img_width * a and vertical shift is
+                randomly sampled in the range 0 < dx < img_height * b
+            shear (float): max abs shear value in degrees between 0 to 180
+            interp: cv2 interpolation method
+            border: cv2 border method
+            border_value: cv2 border value for border=cv2.BORDER_CONSTANT
+        """
+        if scale is not None:
+            assert isinstance(scale, tuple) and len(scale) == 2, \
+                "Argument scale should be a tuple of two floats, e.g (a, b)"
+
+        if translate_frac is not None:
+            assert isinstance(translate_frac, tuple) and len(translate_frac) == 2, \
+                "Argument translate_frac should be a tuple of two floats, e.g (a, b)"
+
+        assert shear >= 0.0, "Argument shear should be between 0.0 and 180.0"
+
+        super(Affine, self).__init__()
+        self._init(locals())
+
+    def _get_augment_params(self, img):
+
+        if self.scale is not None:
+            scale = self._rand_range(self.scale[0], self.scale[1])
+        else:
+            scale = 1.0
+
+        if self.translate_frac is not None:
+            max_dx = self.translate_frac[0] * img.shape[1]
+            max_dy = self.translate_frac[1] * img.shape[0]
+            dx = np.round(self._rand_range(-max_dx, max_dx))
+            dy = np.round(self._rand_range(-max_dy, max_dy))
+        else:
+            dx = 0
+            dy = 0
+
+        if self.shear > 0.0:
+            shear = self._rand_range(-self.shear, self.shear)
+            sin_shear = math.sin(math.radians(shear))
+            cos_shear = math.cos(math.radians(shear))
+        else:
+            sin_shear = 0.0
+            cos_shear = 1.0
+
+        center = (img.shape[1::-1] * np.array((0.5, 0.5))) - 0.5
+        deg = self._rand_range(-self.rotate_max_deg, self.rotate_max_deg)
+
+        transform_matrix = cv2.getRotationMatrix2D(tuple(center), deg, scale)
+
+        # Apply shear :
+        if self.shear > 0.0:
+            m00 = transform_matrix[0, 0]
+            m01 = transform_matrix[0, 1]
+            m10 = transform_matrix[1, 0]
+            m11 = transform_matrix[1, 1]
+            transform_matrix[0, 1] = m01 * cos_shear + m00 * sin_shear
+            transform_matrix[1, 1] = m11 * cos_shear + m10 * sin_shear
+            tx = center[0] * (1.0 - m00) - center[1] * transform_matrix[0, 1]
```

Comment:
The lines 227 and 228:
```
transform_matrix[0, 1] = m01 * cos_shear + m00 * sin_shear
transform_matrix[1, 1] = m11 * cos_shear + m10 * sin_shear
```
inserts shear into rotational part : 
```
[[ scale * cos(a),    scale*sin(a+shear)]
[-scale*sin(a),       scale*cos(a+shear)]]
```
and yes, `tx` and `ty` keep the center unchanged with shear